monai/metrics/metric.py

```diff
@@ -0,0 +1,215 @@
+        else:
+            raise ValueError("y_pred or y must be a list of `channel-first` Tensors or a `batch-first` Tensor.")
+
+        return ret
+
+    def _compute_list(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
+        """
+        Excute the computation for the y_pred and y items of a iteration, the data is in the list shape.
+        Will concat the results to guarantee the output shape of ret is BCHW[D], otherwise it's list of batch-first,
+        which is against our principle that data in metrics should be BCHW[D] or list of channel-first.
+        Note: subclass may enhance the operation with multi-threads to accelerate.
+
+        """
+        ret: TensorOrList
+        if y is not None:
+            ret = [self._compute_tensor(p.detach().unsqueeze(0), y_.detach().unsqueeze(0)) for p, y_ in zip(y_pred, y)]
+        else:
+            ret = [self._compute_tensor(p_.detach().unsqueeze(0), None) for p_ in y_pred]
+        # concat the list of results
+        if isinstance(ret[0], torch.Tensor):
+            ret = torch.cat(ret, dim=0)
+        elif isinstance(ret[0], (list, tuple)) and all([isinstance(i, torch.Tensor) for i in ret[0]]):
+            # if _compute_tensor() returned not only 1 Tensor, concat them separately
+            ret = [torch.cat([k[i] for k in ret], dim=0) for i in range(len(ret[0]))]
+
+        return ret
+
+    @abstractmethod
+    def _compute_tensor(self, y_pred: torch.Tensor, y: Optional[torch.Tensor] = None):
+        """
+        computation logic for the y_pred and y of a iteration, the data should be `batch-first` Tensors.
+        Every subclass metric should implement its own computation logic according to its algorithm.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class Cumulative:
+    """
+    Utility class for the typical cumulative computation process based on PyTorch Tensors.
+    It cumulates tensors in the buffer, then sync across distributed ranks and aggregate.
+    Note: the data list should have the same length every time calling `add()` in a round,
+    it will automatically create buffers according to the length of data list.
+    Execute the steps referring to below examples::
+
+        cum = Cumulative()
+        cum.add(x, y)
+        cum.add(a, b)
+        cum.add(c, d)
+        cum.agrregate()
+        result = cum.get_buffer()
+        cum.reset()
+
+    """
+
+    def __init__(self):
+        self.buffer_num: int = 0
+        self._buffers: Optional[List[List[torch.Tensor]]] = None
+        self._synced_tensors: Optional[List[Optional[torch.Tensor]]] = None
+        self._synced: bool = False
+
+    def reset(self):
+        """
+        Reset the buffers for cumulative tensors and the synced results.
+
+        """
+        self._buffers = None
+        self._synced_tensors = None
+        self._synced = False
+
+    def add(self, *data: torch.Tensor):
```

Comment:
I think every time `self.add` is called, it also should set `self._synced=False`?

Reply:
Makes sense, added in the latest commit.
I think we only need to support 2 regular use cases: (1) add, add, add, aggregate, reset. (2) add, add, aggregate, add, add, aggregate, add, add, ...

Thanks.